src/tree-sitter-language-mode.js

```diff
@@ -947,55 +950,97 @@ class HighlightIterator {
       }
     );
 
-    this.iterators = [
-      this.languageMode.rootLanguageLayer.buildHighlightIterator()
-    ];
-    for (const marker of injectionMarkers) {
-      this.iterators.push(marker.languageLayer.buildHighlightIterator());
-    }
-    this.iterators.sort((a, b) => b.getIndex() - a.getIndex());
-
     const containingTags = [];
     const containingTagStartIndices = [];
     const targetIndex = this.languageMode.buffer.characterIndexForPosition(
       targetPosition
     );
-    for (let i = this.iterators.length - 1; i >= 0; i--) {
-      this.iterators[i].seek(
-        targetIndex,
-        containingTags,
-        containingTagStartIndices
-      );
+
+    this.iterators = [];
+    const iterator = this.languageMode.rootLanguageLayer.buildHighlightIterator();
+    if (iterator.seek(targetIndex, containingTags, containingTagStartIndices)) {
+      this.iterators.push(iterator);
     }
-    this.iterators.sort((a, b) => b.getIndex() - a.getIndex());
+    for (const marker of injectionMarkers) {
+      const iterator = marker.languageLayer.buildHighlightIterator();
+      if (
+        iterator.seek(targetIndex, containingTags, containingTagStartIndices)
+      ) {
+        this.iterators.push(iterator);
+      }
+    }
+    this.iterators.sort((a, b) => b.compare(a));
     return containingTags;
   }
 
   moveToSuccessor() {
-    const lastIndex = this.iterators.length - 1;
-    const leader = this.iterators[lastIndex];
-    leader.moveToSuccessor();
-    const leaderCharIndex = leader.getIndex();
-    let i = lastIndex;
-    while (i > 0 && this.iterators[i - 1].getIndex() < leaderCharIndex) i--;
-    if (i < lastIndex) this.iterators.splice(i, 0, this.iterators.pop());
+    let leader = last(this.iterators);
+    if (leader.moveToSuccessor()) {
+      const leaderIndex = this.iterators.length - 1;
+      let i = leaderIndex;
+      while (i > 0 && this.iterators[i - 1].compare(leader) < 0) i--;
+      if (i < leaderIndex) {
+        this.iterators.splice(i, 0, this.iterators.pop());
+        this.trackLayerDepth();
+      }
+    } else {
+      this.iterators.pop();
+      this.trackLayerDepth();
+    }
+  }
+
+  trackLayerDepth() {
+    let i = this.iterators.length - 1;
+    let iterator = this.iterators[i];
+    if (!iterator) return;
+    const offset = iterator.getOffset();
+    if (iterator.languageLayer.isOpaque) {
+      this.opaqueLayerDepth = iterator.languageLayer.depth;
+    }
+
+    while (i > 0) {
+      i--;
+      iterator = this.iterators[i];
+      if (iterator.startOffset > offset) break;
+      if (iterator.languageLayer.isOpaque) {
+        const { depth } = iterator.languageLayer;
+        if (depth > this.opaqueLayerDepth) {
+          this.opaqueLayerDepth = depth;
+        }
+      }
+    }
```

Comment:
Whenever the ordering of the iterators changes, we call this method which records the depth of any *opaque* layers that span the current position. By consulting this depth value, we prevent highlights from being emitted from parent layers while we're inside of child layers.

In particular, when we've injected a Rust layer into a token tree node in the parent syntax tree, it prevents us from rendering highlights for the token tree node. That would be bad, because we'd get duplicated scopes, and conflicting scopes for some tokens. For example, identifier in the code would get styles as *both* `type_identifier` (via the injected layer) and regular `identifier` (via the original token tree).